google_guest_agent/events/metadata/metadata.go

```diff
@@ -0,0 +1,80 @@
+//  Copyright 2023 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+// Package metadata implement the metadata events watcher.
+package metadata
+
+import (
+	"context"
+	"net"
+	"net/url"
+	"time"
+
+	"github.com/GoogleCloudPlatform/guest-agent/metadata"
+	"github.com/GoogleCloudPlatform/guest-logging-go/logger"
+)
+
+const (
+	// WatcherID is the metadata watcher's ID.
+	WatcherID = "metadata-watcher"
+)
+
+var (
+	// arbitrarily defined wait duration(keeps behavioral backward compatibility).
+	retryWaitDuration = 5 * time.Second
+)
+
+// Watcher is the metadata event watcher implementation.
+type Watcher struct {
+	client         *metadata.Client
+	failedPrevious bool
+}
+
+// New allocates and initializes a new Watcher.
+func New() *Watcher {
+	return &Watcher{
+		client: metadata.New(),
+	}
+}
+
+// ID returns the metadata event watcher id.
+func (mp *Watcher) ID() string {
+	return WatcherID
+}
+
+// Run listens to metadata changes and report back the event.
+func (mp *Watcher) Run(ctx context.Context) (bool, string, interface{}) {
+	if mp.failedPrevious {
+		time.Sleep(1 * time.Second)
+	}
+
+	descriptor, err := mp.client.Watch(ctx)
+	if err != nil {
+		// Only log error once to avoid transient errors and not to spam the log on network failures.
+		if !mp.failedPrevious {
+			if urlErr, ok := err.(*url.Error); ok {
+				if _, ok := urlErr.Err.(*net.OpError); ok {
+					logger.Errorf("Network error when requesting metadata, make sure your instance has an active network and can reach the metadata server.")
+				}
+			}
+			logger.Errorf("Error watching metadata: %s", err)
+			mp.failedPrevious = true
+		}
+	} else {
+		mp.failedPrevious = false
+	}
+
+	// If mds failed to respond we'll produce an event with a nil event data pointer.
+	return true, WatcherID + ",longpool", descriptor
```

Comment:
if returned event ID (`WatcherID + ",longpool"`) is not expected to be unique we can define this as a constant. 
also the typo `longpoll`.

Reply:
Sounds good. Thanks.

Reply:
done.